### PR TITLE
Fix ORANGE reentrant boundary crossings

### PR DESCRIPTION
**Pull request by sethrj**
When crossing a boundary, ORANGE assumed that it always leaves the volume. However, if the direction changes during a boundary crossing (after moved to the boundary), the track may immediately reenter. We now check for this case by seeing if the dot product of the track's direction and the local surface normal changes sign.

Previously ORANGE failed with the dreaded
```
celeritas: internal assertion failed: init.volume
```
one surface crossing *after* it entered the wrong volume (potentially many steps later). Now it correctly leaves the volume alone and continues on its way.

Interestingly now the first difference between a detailed track history in VecGeom and ORANGE in the heuristic test (track 34 for three spheres) is where VecGeom gets "stuck" for a few bumps at the point where ORANGE used to move to the wrong volume.

Vecgeom:
```
 - move_to_boundary ==> {-0.11703624748274, 1.5587272149883, -2.5605999269761} (r = 3), SurfaceId{18446744073709551615}
 - set_dir (mu=0.98372053567654) ==> {0.63091932842407, 0.653146579347, 0.41873660803449}dot_norm = -0.0426600033379
 - cross_boundary from VolumeId{1} ==> VolumeId{1}
 - ...
 - - find_next_step(0.24702084599404) ==> 0.16272464176204
 - move_to_boundary ==> {0.044453876456372, 1.7259066265188, -2.4534200963974} (r = 3), SurfaceId{}
 - cross_boundary from VolumeId{1} ==> VolumeId{1}
 - find_next_step(4.4478322229285e-06) ==> 1e-08
 - move_to_boundary ==> {0.044453882765565, 1.7259066330503, -2.45342009221} (r = 3.0000000004266), SurfaceId{}
 - cross_boundary from VolumeId{1} ==> VolumeId{1}
 - find_next_step(0.0014527008593702) ==> 1e-08
 - move_to_boundary ==> {0.044453889074759, 1.7259066395818, -2.4534200880226} (r = 3.0000000008532), SurfaceId{}
 - cross_boundary from VolumeId{1} ==> VolumeId{2}
```

ORANGE (where the volume IDs are +1 compared to vecgeom):
```
 - move_to_boundary ==> {-0.11703624748274, 1.5587272149883, -2.5605999269761} (r = 3), SurfaceId{2}
 - set_dir (mu=0.98372053567654) ==> {0.63091932842407, 0.653146579347, 0.41873660803449}dot_norm = -0.0426600033379
 - cross_boundary from VolumeId{2} ==> VolumeId{2}
 - ...
 - find_next_step(0.24702084599404) ==> 0.16272464176203
 - move_to_boundary ==> {0.044453876456368, 1.7259066265188, -2.4534200963974} (r = 3), SurfaceId{2}
 - cross_boundary from VolumeId{2} ==> VolumeId{3}
 - find_next_step(4.4478322229285e-06) ==> 4.4478322229285e-06
```

Disappointingly, though, the magnetic field testem3 still fails with ORANGE with a similar init.volume error.

### CODE REVIEW

**Review by amandalund — APPROVED** [review]
Yay, one more down! Kind of similar to the problem we ran into in #354 where one of the secondary gammas from positron annihilation on a boundary moved back into the original volume. Thanks for tracking this down and for the fix @sethrj! I reran some of the problems in #460 and a few more are passing now :)

File: src/orange/OrangeTrackView.hh [hunk 1]
```diff
@@ -419,12 +433,40 @@ CELER_FUNCTION void OrangeTrackView::cross_boundary()
...
+        Real3 normal = tracker.normal(this->pos(), this->surface_id());
+        if (orig_dot_product != (dot_product(normal, this->dir()) >= 0))
+        {
+            // Direction change means we don't cross the boundary! (Or maybe we
+            // already uncrossed it due to mag field and are re-crossing due to
+            // some other reason?)
+            states_.boundary[thread_]
+                = flip_boundary(states_.boundary[thread_]);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/orange/OrangeTrackView.hh:468]: Yep, nice catch!

File: src/orange/Data.hh [hunk 2]
```diff
@@ -288,6 +291,7 @@ inline void resize(OrangeStateData<Ownership::value, M>* data,
     resize(&data->vol, size);
     resize(&data->surf, size);
     resize(&data->sense, size);
+    resize(&data->boundary, size);
```
[the remark this answers is not in the document]
Comment (sethrj) [reply src/orange/Data.hh:294]: We shouldn't need to: it's conditionally queried based on whether a track is on the boundary, which only happens after move-to-boundary in which case we set it.

[the remark this answers is not in the document]
Comment (amandalund) [reply src/orange/Data.hh:294]: Makes sense, the only exception I could think of is if a track is initialized on a boundary, which isn't allowed in orange anyway.

**Pull request merged**